Register dispute agents in apitest arbitrator daemon

ghubstan opened the pull request:
This PR adds a new `CoreDisputeAgentService` for handling dispute agent registration in arbitrator daemons.  API methods for dispute agent registration in `mainnet` mode are out of scope of this PR;  these changes are for supporting the auto-registration of a mediator and refund agent in `regtest/dao` mode for `apitest` cases.  

Some minor, unrelated changes are included, and listed below.

 * Inject new `CoreDisputeAgentService` into `CoreApi`.

 * Change a `DisputeAgentManager` 'dispute agent saved' log statement to indicate type of agent being saved.

 * Add `import static bisq.common.app.Version.VERSION` to `CoreApi`, and add `@SuppressWarnings("SameReturnValue")` annotation to the `getVersion` method.

 * Fix comment in `Scaffold` class.

[PR 4461](https://github.com/bisq-network/bisq/pull/4461) should be reviewed/merged before this one.

ghubstan committed:
Make CoreApi a singleton

ghubstan committed:
Add @VisibleForTesting annotation

The access modifier on CompositeOptionSet was changed to public
in commit c402386, to make it visible to the apitest config
class.  The annotation added in this change concisely explains
why the class is not package protected.

ghubstan committed:
Remove commented code from BisqAppMain

Also fixed line length of a comment, and removed two extra white lines.

ghubstan committed:
Register dispute agents in apitest arbitrator daemon

This PR adds a new CoreDisputeAgentService for handling dispute agent
registration in arbitrator daemons.  API methods for dispute agent
registration in mainnet mode are out of scope of this PR;  these
changes are for supporting the auto-registration of a mediator and
refund agent in regtest/dao mode for apitest cases.

Some minor, unrelated changes are included.

 * Inject new CoreDisputeAgentService into CoreApi.

 * Change a DisputeAgentManager 'dispute agent saved' log statement
   to indicate type of agent being saved.

 * Add import static bisq.common.app.Version.VERSION to CoreApi, and
   add @SuppressWarnings("SameReturnValue") annotation to the getVersion
   method.

 * Fix comment in Scaffold class.

ghubstan committed:
Add empty comments to keep codacy happy

ghubstan commented:
The codacy issue `Perhaps 'coreDisputeAgentService' could be replaced by a local variable.` won't go away until API methods for registering dispute agents are implemented.   

The field declaration needs to stay as it is for auto-registering dispute agents in test arbitration daemons.

sqrrm commented:
@ghubstan @ripcurlx This PR can't be merged since codacy complains, and the complaint is legit if this were to be the end state of this work, but since it's a half way state, what do we do about it?

ghubstan committed:
Fix codacy issue caused by unused class level var

Commit 40dd1a9 @Injects a new CoreDisputeAgentService into CoreApi, but
it was unused by CoreApi, resulting in a legit codacy issue: "Perhaps
'coreDisputeAgentService' could be replaced by a local variable."

This change adds two pass through methods to CoreApi that use
the new CoreDisputeAgentService.  It should fix the codacy problem.
These new pass through methods are not useful yet, and may be
removed in a future PR for registering dispute agents with the API,
but CoreDisputeAgentService is needed now for apitest cases.

The CoreDisputeAgentService instance adds a listener to the
P2PService that tells it when the p2p network is bootstrapped (upon
receiving an onDataReceived event), then it automatically registers
dispute agents (if the daemon is an arbitrator node running in
regtest/dao mode).

sqrrm commented on core/src/main/java/bisq/core/api/CoreDisputeAgentService.java:
```diff
@@ -0,0 +1,187 @@
+        return refundAgentManager.getDisputeAgentByNodeAddress(nodeAddress);
+    }
+
+    private boolean shouldRegisterTestArbitrationDisputeAgents() {
```
This seems odd, having all this test code in a non test class. Seems like part of this is useful for mainnet as well.

ghubstan replied:
Until I write the API methods for registering dispute agents, it can only be done in the UI.  The test harness starts up supporting apps (bitcoind,seednode,arbnode,etc) in separate linux processes, so test cases can only change state in those supporting Bisq apps via the gRPC API.  I'm doing a little triage here, and writing gRPC API support for registering dispute agents from a CLI was moved to the back of the line.  Maybe that's a mistake, but this is my reasoning...

This odd class was a short cut for automatically registering dispute agents for api test cases only.  I did not want to automatically register dispute agents in any daemon but a test arb node.  On mainnet, a user should use the CLI to do this; during testing, a dev (or Travis CI) should not have to run a UI to register dispute agents during the dev/test cycle.

This is a short cut to getting a "simplest trading script" ready for demonstration -- but well tested, and without having to do anything in a UI.  I did not think that an arbitrator CLI + daemon would be in the scope of the "simplest trading script" work (maybe that was a mistake), and I intended to remove this test related code when I do build out the API to support arbitrator CLI + daemons.

It took much more time figuring out when dispute agents can be registered than coding this class.  If I had known already, I could have thought this through better than I did, and built the gRPC API support for registering dispute agents from a CLI instead of this temporary short cut.  If it's not acceptable, I can build the API support next, and remove this test code.

sqrrm replied:
I don't think we should allow mainnet registration of dispute agent via cli for now. There are a lot of very specific dispute handling that I think is handled on the UI side. A UI of some sort would also be needed since it's a human facing role that's done manually.

For testing it's fine to have an automatically registered agent of course. Still would prefer if the test specifics were moved to the tests package.

ghubstan replied:
How about a new API method `registerdisputeagent [dispute-agent-type]` that does not work on mainnet?  It would throw an exception to a mainnet CLI, and the user would see an `Error: dispute agents must be registered in the Bisq UI` msg.

Headless regtest api test cases could use this method.

Sound OK?

sqrrm replied:
Yeah, sounds better. Let's go with that.

ghubstan replied:
The commits are a fix for the weird short cut I took for automatically registering dispute agents in api test cases.  Now it's done over gRPC with a `registerdisputeagent` method that will not work on mainnet.

It required some refactoring in the test harness so api calls could be made to an arbitration daemon, and there are a few other minor changes:  renaming a class, line formatting in the cli, and removing hard coded "localhost" strings.

ghubstan replied:
I have an unresolved file conflict, might need some help tomorrow.

sqrrm reviewed (commented):
I might not get what's going on, but why are there test methods in CoreDisputeAgentService.java?

ghubstan committed:
Add DisputeAgents service proto def

Also moved GetVersion proto def to bottom of file (for alphabetic ordering).

ghubstan committed:
Add GrpcDisputeAgentsService boilerplate

ghubstan committed:
Add GrpcDisputeAgentsService to server

ghubstan committed:
Create disputeAgentsService instance in GrpcStubs

ghubstan committed:
Implement the new registerdisputeagent api method

This method cannot be used on mainnet, it is to support auto
registration of mediators and refund agents in api test cases.
It also does not support registration of test arbitrators -- not
needed for api's support for simple trading functionality.

ghubstan committed:
Add registerdisputeagent method to cli

This method cannot be used on mainnet, and is intended for developer
use only.

Some formatting was done to shorten line lengths.

ghubstan committed:
Refactor apitest to support gRPC calls to any daemon

Api test cases can now make api calls to arbitrator, bob & alice
daemons, not just alice's daemon.

ghubstan committed:
Delete renamed class

ghubstan committed:
Add RegisterDisputeAgentsTest

ghubstan committed:
Fix import statements

ghubstan committed:
Resolve merge conflict

ghubstan committed:
Revert "Resolve merge conflict"

This reverts commit c5639e694ffa62a4d23ec8d5c6db86647a343d33,
which failed to resolve the file conflict.

ghubstan committed:
Downgrade jfoenix dependency

The jfoenix upgrade in PR 4443 was reverted in PR 4463 due to
incompatibility with JDK 10.  This change avoids an accidental
re-upgrade during merge.

ghubstan commented:
Closing, a cleaner PR will be created from a branch based off master, with file conflict resolved.
